### PR TITLE
dbeaver/dbeaver#14620 auto-find 'local clients' on Linux and macOS on AArch64

### LINKED ISSUE #14620: Find local clients on Linux & macOS under AArch64
We already do that on Windows and [macOS on x86_64](https://github.com/dbeaver/dbeaver/issues/7976). Now it's time to do the same on Linux and on macOS under AArch64.

It's a standard practice to add a symlink to `/usr/bin/` for pretty much all the binaries that we need to find. Using such symlinks, we can discover the location of local clients. We can also try to locate something in such folders as `/etc/alternatives/`.

### PR DESCRIPTION
**Pull request by flambchap**
Please check:
- [ ] Detection of PG native clients on Intel-based Macs (homebrew, EDB installers, Postgres.app)
- [x] Detection of PG native clients on AArch64 Macs (homebrew, EDB installers)
- [ ] Detection of MySQL native clients on Intel-based Macs with homebrew
- [ ] Detection of MariaDB native clients on Intel-based Macs with homebrew
- [x] Detection of MySQL native clients on AArch64 Macs with homebrew
- [x] Detection of MariaDB native clients on AArch64 Macs with homebrew
- [x] Detection of PG packages on Linux
- [x] Detection of MySQL packages on Linux
- [x] Detection of MariaDB packages on Linux
- [x] Detection of PG installations on Windows
- [x] Detection of MySQL installations on Windows
- [x] Detection of MariaDB installations on Windows
- [x] Performance of the New Connection wizard for any of the affected dbs. Make sure that the Main tab opens reasonably fast and that the Local Client combo box is responsive.

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/PostgreDataSourceProvider.java [hunk 1]
```diff
@@ -183,60 +189,53 @@ public synchronized static void findLocalClients() {
...
+            if (RuntimeUtils.isOSArchAArch64()) {
+                foldersToExamine.add("/opt/homebrew/bin");
+                foldersToExamine.add("/opt/homebrew/Cellar");
+                foldersToExamine.add("/opt/homebrew/opt");
+            }
+        }
+
+        for (String folder : foldersToExamine) {
+            Path folderPath = Path.of(folder);
+            if (!folderPath.toFile().exists()) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/PostgreDataSourceProvider.java]: `Files.notExists`

File: bundles/org.jkiss.utils/src/org/jkiss/utils/IOUtils.java [hunk 2]
```diff
@@ -441,4 +441,18 @@ public static boolean isLocalPath(Path filePath) {
...
     }
 
+    /**
+     * Returns the <em>grandparent path</em>, or {@code null} if this path does not have a grandparent.
+     *
+     * @param path a path
+     * @return a path representing the path's grandparent
+     */
+    @Nullable
+    public static Path getGrandparent(@NotNull Path path) {
```
Comment (ShadelessFox) [inline bundles/org.jkiss.utils/src/org/jkiss/utils/IOUtils.java]: This method seems very specific and only used in a single place.

File: plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/PostgreDataSourceProvider.java [hunk 3]
```diff
@@ -183,60 +189,53 @@ public synchronized static void findLocalClients() {
...
+        Collection<String> foldersToExamine = new ArrayList<>();
+        foldersToExamine.add("/usr/bin");
+        foldersToExamine.add("/usr/local/bin");
+        if (RuntimeUtils.isLinux()) {
+            foldersToExamine.add("/etc/alternatives");
         }
-        if (fullVersion.length() < 7) {
-            log.warn("Unable to figure out PostgreSQL server branding from psql output!");
-            return "";
+        if (RuntimeUtils.isMacOS()) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/PostgreDataSourceProvider.java]: Shouldn't it be rather an `else if` statement?

File: bundles/org.jkiss.utils/src/org/jkiss/utils/IOUtils.java (hunk 2 above)
Comment (flambchap) [reply bundles/org.jkiss.utils/src/org/jkiss/utils/IOUtils.java]: I'll reuse it in a different bundle a bit later (PR is not ready yet)

**Comment by E1izabeth** [comment]
Is it possible to suggest downloading of the clients if the path was not detected? Like with drivers

**Comment by flambchap** [comment]
>Is it possible to suggest downloading of the clients if the path was not detected? Like with drivers

We do that on Windows. It requires hosting these tools on our website. We can do that for other platforms, but it requires quite a bit of work. I'll create a separate ticket for that.

File: bundles/org.jkiss.utils/src/org/jkiss/utils/CommonUtils.java [hunk 4]
```diff
@@ -181,8 +181,12 @@ public static boolean isEmpty(@Nullable Map<?, ?> value) {
     }
 
     @Nullable
-    public static <T> T getFirstOrNull(@NotNull List<T> list) {
-        return list.isEmpty() ? null : list.get(0);
+    public static <T> T getFirstOrNull(@NotNull Iterable<T> iterable) {
+        Iterator<T> iterator = iterable.iterator();
+        if (iterator.hasNext()) {
+            return iterator.next();
+        }
```
Comment (ShadelessFox) [inline bundles/org.jkiss.utils/src/org/jkiss/utils/CommonUtils.java:188]:
```suggestion
        for (T element : iterable) {
            return element;
        }
```

Okay, I'm joking

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/connection/NativeClientLocationUtils.java [hunk 5]
```diff
@@ -16,47 +16,116 @@
...
 
-    public static File[] getSubdirectories(File... dirs) {
-        return getStreamOfSubdirectories(dirs).toArray(File[]::new);
+    /**
+     * Finds native client locations. Designed to work on Unix and Unix-like systems.
+     * It does so by examining some common disk locations for the clients. A user of this function
+     * can supply extra folders to examine.
+     *
+     * @param extraFoldersToExamine extra folders to look for the clients
+     * @param fileEndings the endings of full file paths for search for
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/connection/NativeClientLocationUtils.java]:
```suggestion
     * @param fileEndings the endings of full file paths to search for
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/connection/NativeClientLocationUtils.java [hunk 6]
```diff
@@ -16,47 +16,116 @@
...
+                            Path grandparent = getGrandparent(file);
+                            if (grandparent != null) {
+                                result.put(grandparent.toString(), grandparentPathToClientLocationMapper.apply(grandparent));
+                            }
+                        }
+                        return FileVisitResult.SKIP_SIBLINGS;
+                    }
+                });
+            } catch (IOException e) {
+                log.warn(String.format("Unable to examine folder %s while looking for a client home", folder), e);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/connection/NativeClientLocationUtils.java]: Can use modern `"lol %s".formatted("yes")`

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/connection/NativeClientLocationUtils.java [hunk 7]
```diff
@@ -16,47 +16,116 @@
...
+                continue;
+            }
+            try {
+                Files.walkFileTree(folderPath, new SimpleFileVisitor<>() {
+                    @Override
+                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+                        if (!somethingEndsWith(file, fileEndings)) {
+                            return FileVisitResult.CONTINUE;
+                        }
+                        if (file.toFile().canExecute()) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/connection/NativeClientLocationUtils.java]:
```suggestion
                        if (Files.isExecutable(file)) {
```

**Review by E1izabeth — APPROVED** [review]

**Review by ShadelessFox — APPROVED** [review]

**Pull request merged**